_posts/2020-04-13-Julia-Antipatterns.md

```diff
@@ -0,0 +1,831 @@
+Using an immutable data structure avoids that.
+
+Here is example for a `Dict`.
+It highlights that while the time taken to get a value is expected O(1), i.e. constant-time,
+that constant is not tiny as it needs to compute the `hash`.
+`hash` for `Symbol` is fairly fast, for `String` is a  bit larger.
+For more complicated objects it can be quiet large, often larger than `isequal` as designing a good `hash` function is hard.
+But the case we see in this antipattern are mostly `Dict{Symbol}` or `Dict{String}`
+
+**Input:**
+
+<div class="jupyter-input jupyter-cell">
+
+```julia
+dict = Dict([:a=>1, :b=>2, :c=>3, :d=>4, :e=>5])
+@btime $(dict)[:d];
+```
+</div>
+
+**Output:**
+
+<div class="jupyter-stream jupyter-cell">
+
+```
+
+  8.456 ns (0 allocations: 0 bytes)
+
+```
+</div>
+
+**Input:**
+
+<div class="jupyter-input jupyter-cell">
+
+```julia
+str_dict = Dict(string(k)=> v for (k,v) in dict)  # convert all the keys to strings
+@btime $(str_dict)["d"];
+```
+</div>
+
+**Output:**
+
+<div class="jupyter-stream jupyter-cell">
+
+```
+
+  18.077 ns (0 allocations: 0 bytes)
+
+```
+</div>
+
+One alternative is [OrderedCollection's](https://github.com/JuliaCollections/OrderedCollections.jl/) `LittleDict`.
+This is a naive pair of lists based dictionary, with expected time `O(n)` but with a much lower cost per actual element as it doesn't need to `hash` anything.
+
+**Input:**
+
+<div class="jupyter-input jupyter-cell">
+
+```julia
+using OrderedCollections
+little_dict = LittleDict(dict)
+@btime $(little_dict)[:d];
+```
+</div>
+
+**Output:**
+
+<div class="jupyter-stream jupyter-cell">
+
+```
+
+  4.793 ns (0 allocations: 0 bytes)
+
+```
+</div>
+
+It also comes in a immutable form via `freeze` (or by constructing it with `Tuple`s)
+
+**Input:**
+
+<div class="jupyter-input jupyter-cell">
+
+```julia
+frozen_little_dict = freeze(LittleDict(dict))
+@btime $(frozen_little_dict)[:d];
+```
+</div>
+
+**Output:**
+
+<div class="jupyter-stream jupyter-cell">
+
+```
+
+  4.547 ns (0 allocations: 0 bytes)
+
+```
+</div>
+
+But the real winner here is the `NamedTuple`.
+Which performs [constant-folding](https://en.wikipedia.org/wiki/Constant_folding) to remove the indexing operation entirely, and just put the result directly into the AST during compilation.
+
+**Input:**
+
+<div class="jupyter-input jupyter-cell">
+
+```julia
+named_tuple = (; dict...)  # create a NamedTuple with the same content as the dict
+@btime $(named_tuple)[:d];
+```
+</div>
+
+**Output:**
+
+<div class="jupyter-stream jupyter-cell">
+
+```
+
+  0.032 ns (0 allocations: 0 bytes)
+
+```
+</div>
+
+If one is thinking _"I have some constants and I want to group them"_ then look no further than the `NamedTuple`.
+It is the right answer.
+`Dict` is best when you don't know all the keys when you write the code, and/or if values need to be updated.
+
+## Over constraining argument types
+
+Type constraints in Julia are **only** for dispatch.
```

Comment:
Maybe phrase in some way that this is an opinion. From a rhetoric perspective, the claim is stronger than the title of the section, which is what you try to prove., so starting with it may put off the reader

Reply:
Title is vague, any constraint beyond what is needed for dispatch is an over-contraint.
But i have added some hedging,
and a disclaimer at the top:

To quote the introduction to Press, Teukolsky, Vetting and Flannery's "Numerical Recipes":

_We do, therefore, offer you our practical judgements wherever we can.
As you gain experience, you will form your own opinion of how reliable our advice is.
Be assured that it is not perfect!_